packages/eslint-plugin/tests/rules/no-misused-spread.test.ts

```diff
@@ -0,0 +1,904 @@
+import { RuleTester } from '@typescript-eslint/rule-tester';
+
+import rule from '../../src/rules/no-misused-spread';
+import { getFixturesRootDir } from '../RuleTester';
+
+const rootPath = getFixturesRootDir();
+
+const ruleTester = new RuleTester({
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    tsconfigRootDir: rootPath,
+    project: './tsconfig.json',
+  },
+});
+
+ruleTester.run('no-misused-spread', rule, {
+  valid: [
+    'const a = [...[1, 2, 3]];',
+    'const a = [...([1, 2, 3] as const)];',
+
+    `
+      declare const data: any;
+      const a = [...data];
+    `,
+
+    `
+      declare const data: number[] | any;
+      const a = [...data];
+    `,
+
+    `
+      declare const data: number[] & any;
+      const a = [...data];
+    `,
+
+    `
+      const a = [1, 2, 3];
+      const b = [...a];
+    `,
+
+    `
+      const a = [1, 2, 3] as const;
+      const b = [...a];
+    `,
+
+    `
+      declare function getArray(): number[];
+      const a = [...getArray()];
+    `,
+
+    `
+      declare function getTuple(): readonly number[];
+      const a = [...getTuple()];
+    `,
+
+    `
+      const iterator = {
+        *[Symbol.iterator]() {
+          yield 1;
+          yield 2;
+          yield 3;
+        },
+      };
+
+      const a = [...iterator];
+    `,
+
+    `
+      declare const data: Iterable<number> | number[];
+
+      const a = [...data];
+    `,
+
+    `
+      declare const data: Iterable<number> & number[];
+
+      const a = [...data];
+    `,
+
+    `
+      declare function getIterable(): Iterable<number>;
+
+      const a = [...getIterable()];
+    `,
+
+    'const o = { ...{ a: 1, b: 2 } };',
+
+    'const o = { ...({ a: 1, b: 2 } as const) };',
+
+    `
+      declare const obj: any;
+
+      const o = { ...obj };
+    `,
+
+    `
+      declare const obj: { a: number; b: number } | any;
+
+      const o = { ...obj };
+    `,
+
+    `
+      declare const obj: { a: number; b: number } & any;
+
+      const o = { ...obj };
+    `,
+
+    `
+      const obj = { a: 1, b: 2 };
+      const o = { ...obj };
+    `,
+
+    `
+      declare const obj: { a: number; b: number };
+      const o = { ...obj };
+    `,
+
+    `
+      declare function getObject(): { a: number; b: number };
+      const o = { ...getObject() };
+    `,
+
+    `
+      class A {
+        a = 1;
+        b = 2;
+      }
+
+      const o = { ...new A() };
+    `,
+
+    `
+      class A {
+        a = 1;
+        b = 2;
+      }
+
+      const a = new A();
+
+      const o = { ...a };
+    `,
+
+    `
+      function f() {}
+
+      f.prop = 1;
+
+      const o = { ...f };
+    `,
+
+    `
+      const f = () => {};
+
+      f.prop = 1;
+
+      const o = { ...f };
+    `,
+
+    `
+      function* generator() {}
+
+      generator.prop = 1;
+
+      const o = { ...generator };
+    `,
+
+    `
+      declare const promise: Promise<number>;
```

Comment:
[Bug] Err, wouldn't we want the rule to flag on this case? And on the `{ ...new A() }` case? ...and now that we're looking at it, `{ ...(class A { static value = 1 }) }` and other `static`-y things?

As a prospective user of this rule, I would have expected instances of classes to not be allowed. Am I totally off base here?

My hunch is that this is a discovered edge case that should have an option. WDYT?

Reply:
IMO, spreading a class instance is weird and probably an edge case as you're saying.

Not sure about spreading a class declaration though... it looks too obscure. 
Is there anyone who actually does that? 😅

Reply:
Done